File: fastify_server/src/core/orchestrator.ts
import type {
  BootstrapConfig,
  HookCollector,
  HookFn,
  LoaderReports,
  ResolvedBootstrapConfig,
} from "../contract/index.js";
import { mergeConfig } from "../contract/index.js";
import type { RuntimeAdapter } from "../adapters/adapter.js";
import type { Addon, Registry } from "../registry/registry.js";
import { Registry as RegistryClass } from "../registry/registry.js";
import { createContext, consoleLogger } from "../registry/ctx.js";
import type { Logger } from "../registry/ctx.js";

export interface SetupOptions {
  baseDir?: string;
  defaults?: BootstrapConfig;
  logger?: Logger;
}

export async function setup<Server>(
  adapter: RuntimeAdapter<Server>,
  addons: Addon<Server>[] | Registry<Server>,
  userConfig: BootstrapConfig,
  opts: SetupOptions = {},
): Promise<Server> {
  const logger = opts.logger ?? consoleLogger;
  const baseDir = opts.baseDir ?? process.cwd();
  const defaults = opts.defaults ?? {};

  const config: ResolvedBootstrapConfig = mergeConfig(defaults, userConfig, baseDir);

  const registry = addons instanceof RegistryClass ? addons : toRegistry(addons);

  const server = await adapter.createServer(config);

  adapter.decorate(server, "_config", config);

  const hooks: HookCollector<Server> = { init: [], startup: [], shutdown: [] };
  const ctx = createContext(adapter, config, hooks, logger);

  const reports: LoaderReports = await registry.runAll(server, config, ctx);
  adapter.decorate(server, "_loaderReports", reports);

  for (const name of Object.keys(reports)) {
    const r = reports[name]!;
    if (r.errors.length > 0) {
      logger.warn(
        `addon ${name}: ${r.errors.length} error(s); discovered=${r.discovered} registered=${r.registered}`,
      );
      for (const [i, e] of r.errors.entries()) {
        const where = e.path ? ` ${e.path}` : "";
        logger.error(`  addon ${name} error[${i}] step=${e.step}${where}: ${e.error}`);
      }
    } else {
      logger.info(
        `addon ${name}: discovered=${r.discovered} registered=${r.registered} skipped=${r.skipped}`,
      );
    }
  }

  if (config.initial_state && Object.keys(config.initial_state).length > 0) {
    adapter.attachRequestState(server, config.initial_state);
  }

  await runInitHooks(server, config, hooks.init, logger);

  adapter.scheduleHooks(server, { startup: hooks.startup, shutdown: hooks.shutdown }, config);
  adapter.registerGracefulShutdown(server);

  return server;
}

function toRegistry<Server>(addons: Addon<Server>[]): Registry<Server> {
  const reg = new RegistryClass<Server>();
  for (const a of addons) reg.register(a);
  return reg;
}

async function runInitHooks<Server>(
  server: Server,
  config: ResolvedBootstrapConfig,
  inits: HookFn<Server>[],
  logger: Logger,
): Promise<void> {
  for (const fn of inits) {
    const label = (fn as { name?: string }).name || "anonymous";
    try {
      await fn(server, config);
      logger.debug(`init hook '${label}' completed`);
    } catch (err) {
      logger.error(`init hook '${label}' failed:`, err);
    }
  }
}

File: fastapi_server/fastapi_server/bootstrap/core/orchestrator.py
from __future__ import annotations

import inspect
import os
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union

from fastapi_server.bootstrap.contract import (
    BootstrapConfig,
    HookCollector,
    LoaderReport,
    ResolvedBootstrapConfig,
    merge_config,
)
from fastapi_server.bootstrap.registry.ctx import Logger, StdoutLogger, create_context
from fastapi_server.bootstrap.registry.registry import Addon, Registry


@dataclass
class SetupOptions:
    base_dir: Optional[str] = None
    defaults: Dict[str, Any] = field(default_factory=dict)
    logger: Optional[Logger] = None


async def setup(
    adapter: Any,
    addons: Union[List[Addon], Registry],
    user_config: Union[Dict[str, Any], BootstrapConfig],
    opts: Optional[SetupOptions] = None,
) -> Any:
    opts = opts or SetupOptions()
    logger: Logger = opts.logger or StdoutLogger()
    base_dir = opts.base_dir or os.getcwd()

    config: ResolvedBootstrapConfig = merge_config(opts.defaults or {}, user_config, base_dir)

    registry = addons if isinstance(addons, Registry) else _to_registry(addons)

    server = adapter.create_server(config)
    if inspect.isawaitable(server):
        server = await server

    adapter.decorate(server, "_config", config)

    hooks = HookCollector()
    ctx = create_context(adapter, config, hooks, logger)

    reports: Dict[str, LoaderReport] = await registry.run_all(server, config, ctx)
    adapter.decorate(server, "_loader_reports", reports)

    for name, r in reports.items():
        if r.errors:
            logger.warn(
                f"addon {name}: {len(r.errors)} error(s); discovered={r.discovered} registered={r.registered}"
            )
            for i, e in enumerate(r.errors):
                where = f" {e.path}" if e.path else ""
                logger.error(f"  addon {name} error[{i}] step={e.step}{where}: {e.error}")
        else:
            logger.info(
                f"addon {name}: discovered={r.discovered} registered={r.registered} skipped={r.skipped}"
            )

    if config.initial_state:
        adapter.attach_request_state(server, config.initial_state)

    await _run_init_hooks(server, config, hooks.init, logger)

    adapter.schedule_hooks(server, {"startup": hooks.startup, "shutdown": hooks.shutdown}, config)
    adapter.register_graceful_shutdown(server)

    return server


def _to_registry(addons: List[Addon]) -> Registry:
    reg = Registry()
    for a in addons:
        reg.register(a)
    return reg


async def _run_init_hooks(server: Any, config: ResolvedBootstrapConfig, inits: list, logger: Logger) -> None:
    for fn in inits:
        label = getattr(fn, "__name__", "anonymous")
        try:
            result = fn(server, config)
            if inspect.isawaitable(result):
                await result
            logger.debug(f"init hook '{label}' completed")
        except Exception as err:  # noqa: BLE001
            logger.error(f"init hook '{label}' failed:", err)
